core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java

```diff
@@ -467,18 +477,21 @@ public void run() {
                         }
                         boolean success = false;
                         try {
-                            // connect to the node, see if we manage to do it, if not, bail
-                            if (!nodeFoundByAddress) {
-                                logger.trace("[{}] connecting (light) to {}", sendPingsHandler.id(), finalNodeToSend);
-                                transportService.connectToNodeAndHandshake(finalNodeToSend, timeout.getMillis());
+                            Transport.Connection connection;
+                            if (nodeFoundByAddress && transportService.nodeConnected(finalNodeToSend)) {
```

Comment:
@ywelsch these are valid points. Seeing the PR already made me realize it's time to re-write this class as it has grown too tangled. I hoped I can do it after @s1monw pushed this PR in, but I'm not sure it's worth it anymore to push through and fix all of this.

@s1monw how about you push just the transport layer changes and I will use them in a follow up PR to clean up this class?

Re node connections - 

You are right too that this is not good and not optimal. Since the nodes was never removed from the cluster state, it was also never removed from `NodeConnectionService` (I *think*). IMO this is just too complex (and there is nodeFD and node joins that play a role here too, on the master). I suggest we change `NodeConnectionService` to always work on the `ClusterState.nodes()` directly rather than trying to be overly heroic and use the deltas. This will give a hard validation of every CS. This should be very cheep tests. We can do it as another follow - potentially before we do the unicast zen ping work.

How does that sound?